spec/FilesController.spec.js

```diff
@@ -21,11 +21,14 @@ const mockAdapter = {
 
 // Small additional tests to improve overall coverage
 describe('FilesController', () => {
-  it('should properly expand objects', done => {
+  it('should properly expand objects with sync getFileLocation', async () => {
     const config = Config.get(Parse.applicationId);
     const gridFSAdapter = new GridFSBucketAdapter('mongodb://localhost:27017/parse');
+    gridFSAdapter.getFileLocation = (config, filename) => {
+      return config.mount + '/files/' + config.applicationId + '/' + encodeURIComponent(filename);
+    }
     const filesController = new FilesController(gridFSAdapter);
-    const result = filesController.expandFilesInObject(config, function () {});
+    const result = await filesController.expandFilesInObject(config, function () { });
```

Comment:
To make sure it all still works, we shouldn't touch any existing tests but only add new ones. After all, this is a feature addition, so any new tests for the new feature (asynchronous invocation) should be needed. The exception being tests for internal methods. Could you revert changes to existing tests?

Reply:
The expandFilesInObject function depends on getFileLocation, which now supports asynchronous operations. Given this change, I considered two approaches:

1. Check if getFileLocation returns a promise: This would involve adding conditional logic to handle both synchronous and asynchronous scenarios. However, to maintain consistency, this conditional logic would need to be propagated throughout the entire codebase wherever expandFilesInObject is used. This approach seems overly complex and error-prone.

2. Make expandFilesInObject fully async: By making expandFilesInObject an async function and using await for getFileLocation, we simplify the code and ensure it is consistently handling asynchronous operations. This also means updating any functions that rely on expandFilesInObject to be async as well, which I have done, along with adjusting the relevant tests.

I chose the second approach because it results in cleaner, more maintainable code without the need for extensive conditional logic across the codebase. I understand the concern about modifying existing tests, but given the changes, these updates are necessary to ensure that the tests accurately reflect the new asynchronous behavior.

I'm open to further discussion if you have concerns or alternative suggestions.

Reply:
Yes that makes sense. It's good to refactor and optimize internal methods. Just the tests for exposed / public APIs need to stay unchanged.

Reply:
Yeah, as I searched through the project, I only found four instances in three files: UserRouter.js, RestWrite.js, and RestQuery.js. I ensured that all the lines I changed are already covered by existing tests, as confirmed by Codecov.

There was only one function that wasn't already async: RestWrite.expandFilesForExistingObjects. However, that function is only used in the main execute function, which was already async. So, everything should work seamlessly.